provider/source/src/units/categorized_display_name.rs

```diff
@@ -0,0 +1,153 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::collections::HashSet;
+
+use crate::cldr_serde::{self};
+use crate::SourceDataProvider;
+
+use icu::experimental::dimension::provider::units::categorized_display_name::{
+    AreaDisplayNameV1, DurationDisplayNameV1, LengthDisplayNameV1, MassDisplayNameV1,
+    VolumeDisplayNameV1,
+};
+use icu::experimental::dimension::provider::units::display_name::UnitsDisplayName;
+use icu::plurals::PluralElements;
+use icu_provider::prelude::*;
+use icu_provider::DataMarkerAttributes;
+
+macro_rules! impl_categorized_display_name_data_provider {
+    ($category:expr, $display_name:ident) => {
+        impl DataProvider<$display_name> for SourceDataProvider {
+            fn load(&self, req: DataRequest) -> Result<DataResponse<$display_name>, DataError> {
+                self.check_req::<$display_name>(req)?;
+
+                // This category is based on the Marker
+                let category = $category;
+                let (length, unit) = req
+                    .id
+                    .marker_attributes
+                    .split_once('-')
+                    .ok_or_else(|| DataErrorKind::InvalidRequest.into_error())?;
+
+                let units_format_data: &cldr_serde::units::data::Resource = self
+                    .cldr()?
+                    .units()
+                    .read_and_parse(req.id.locale, "units.json")?;
+                let units_format_data = &units_format_data.main.value.units;
+
+                let unit_patterns = match length {
+                    "long" => &units_format_data.long,
+                    "short" => &units_format_data.short,
+                    "narrow" => &units_format_data.narrow,
+                    _ => {
+                        return Err(DataErrorKind::InvalidRequest
+                            .into_error()
+                            .with_debug_context(length))
+                    }
+                }
+                .categories
+                .get(category)
+                .ok_or_else(|| {
+                    DataErrorKind::IdentifierNotFound
+                        .into_error()
+                        .with_debug_context(category)
```

Comment:
Nit: If the category is not found, that's a fairly serious error. It's more likely `InvalidRequest`, not `IdentifierNotFound`.